### ISSUE TITLE
CoreOS 815.0.0/Docker 1.8.2 cannot pull images from docker hub - hangs

### ISSUE THREAD

**Issue by johnt337** [docker/docker#16689]
```
docker version:

Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   fae4436-dirty
 Built:        Thu Sep 24 08:01:15 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   fae4436-dirty
 Built:        Thu Sep 24 08:01:15 UTC 2015
 OS/Arch:      linux/amd64
```

```
docker Info:

Containers: 1
Images: 295
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.2.0-coreos-r1
Operating System: CoreOS 815.0.0
CPUs: 2
Total Memory: 1.958 GiB
Name: docker.vm
ID: WPPL:HA2U:IKEC:C2N7:7SJY:5O47:XMTI:GMHF:ZCMQ:PA4Q:23NZ:MXHG
```

```
uname -a:

Linux docker.vm 4.2.0-coreos-r1 #2 SMP Thu Sep 24 08:00:18 UTC 2015 x86_64 Intel(R) Xeon(R) CPU E5-1650 v2 @ 3.50GHz GenuineIntel GNU/Linux
```

```
Steps to reproduce the problem:

1. docker pull ubuntu
2. docker pull registry
3. docker pull <image>
```

```
Environment details:

OSX 10.10.5 with VMWare Fusion Professional Version 7.1.2 (2779224)
CoreOS 815.0.0 as docker vm
(Darwin projim.local 14.5.0 Darwin Kernel Version 14.5.0: Wed Jul 29 02:26:53 PDT 2015; root:xnu-2782.40.9~1/RELEASE_X86_64 x86_64)
OSX Docker version:
Client:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.5.1
 Git commit:   0a8c2e3
 Built:        Fri Sep 11 01:46:35 UTC 2015
 OS/Arch:      darwin/amd64

Server:
 Version:      1.8.2
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   fae4436-dirty
 Built:        Thu Sep 24 08:01:15 UTC 2015
 OS/Arch:      linux/amd64
```

```
Actual Results:

johnt@docker ~ $ docker pull ubuntu
Using default tag: latest
latest: Pulling from library/ubuntu
d3a1f33e8a5a: Pull complete
c22013c84729: Pull complete
d74508fb6632: Extracting [==================================================>]    683 B/683 B
91e54dfb1179: Layer already being pulled by another client. Waiting.
91e54dfb1179: Layer already being pulled by another client. Waiting.
```

```
Docker daemon logs

Oct 01 03:44:30 docker.vm dockerd[691]: time="2015-10-01T03:44:30.675373326Z" level=info msg="POST /v1.20/images/create?fromImage=ubuntu%3Alatest"
Oct 01 03:44:32 docker.vm dockerd[691]: time="2015-10-01T03:44:32.115229896Z" level=info msg="Image manifest for ubuntu:latest has been verified"
Oct 01 03:44:44 docker.vm dockerd[691]: time="2015-10-01T03:44:44.758692829Z" level=error msg="error copying from layer download progress reader: download canceled"
```

```
After deleting the images and turning debug on for the daemon:

johnt@docker ~ $ docker pull ubuntu
Using default tag: latest
latest: Pulling from library/ubuntu
d3a1f33e8a5a: Pull complete
c22013c84729: Pull complete
d74508fb6632: Pull complete
91e54dfb1179: Layer already being pulled by another client. Waiting.
```

```
Docker daemon logs

Oct 01 03:51:21 docker.vm dockerd[1438]: time="2015-10-01T03:51:21.981865973Z" level=debug msg="Calling POST /images/create"
Oct 01 03:51:21 docker.vm dockerd[1438]: time="2015-10-01T03:51:21.981940265Z" level=info msg="POST /v1.20/images/create?fromImage=ubuntu%3Alatest"
Oct 01 03:51:21 docker.vm dockerd[1438]: time="2015-10-01T03:51:21.982017616Z" level=debug msg="Trying to pull ubuntu from https://registry-1.docker.io v2"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.005598404Z" level=debug msg="Fetched 1 base graphs at 2015-10-01 03:51:22.005580244 +0000 UTC"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.011216184Z" level=debug msg="Reloaded graph with 3 grants expiring at 2017-03-22 19:04:46.713978458 +0000 UTC"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.220205756Z" level=debug msg="Pulling tag from V2 registry: \"latest\""
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.916631378Z" level=debug msg="Key check result: verified"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.916705680Z" level=info msg="Image manifest for ubuntu:latest has been verified"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.917266306Z" level=debug msg="pulling blob \"sha256:bbe1c4256df30171585344c50fab278157cbf2cb3a2016bb720e99e49a759743\" to d3a1f33e8a5a513092f01bb7eb1c2abf4d711e5105390a3fe1ae2248cfde1391"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.917454453Z" level=debug msg="pulling blob \"sha256:911d09728ffd9388b33c62f8ea09d751350a882f719794acb3b186dcf5b958df\" to c22013c8472965aa5b62559f2b540cd440716ef149756e7b958a1b2aba421e87"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.917567797Z" level=debug msg="pulling blob \"sha256:615765bc0d9f82db061f7575d01bd94c99465a6e8dc1a45b793e6fd179e1ddae\" to d74508fb6632491cea586a1fd7d748dfc5274cd6fdfedee309ecdcbc2bf5cb82"
Oct 01 03:51:22 docker.vm dockerd[1438]: time="2015-10-01T03:51:22.917693025Z" level=debug msg="pulling blob \"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4\" to 91e54dfb11794fad694460162bf0cb0a4fa710cfa3f60979c177d920813e267c"
Oct 01 03:51:24 docker.vm dockerd[1438]: time="2015-10-01T03:51:24.005939768Z" level=debug msg="Downloaded c22013c8472965aa5b62559f2b540cd440716ef149756e7b958a1b2aba421e87 to tempfile /var/lib/docker/tmp/GetImageBlob714743613"
Oct 01 03:51:24 docker.vm dockerd[1438]: time="2015-10-01T03:51:24.568914642Z" level=debug msg="Downloaded d74508fb6632491cea586a1fd7d748dfc5274cd6fdfedee309ecdcbc2bf5cb82 to tempfile /var/lib/docker/tmp/GetImageBlob224874360"
Oct 01 03:51:31 docker.vm dockerd[1438]: time="2015-10-01T03:51:31.734185276Z" level=debug msg="Downloaded d3a1f33e8a5a513092f01bb7eb1c2abf4d711e5105390a3fe1ae2248cfde1391 to tempfile /var/lib/docker/tmp/GetImageBlob551874518"
Oct 01 03:51:31 docker.vm dockerd[1438]: time="2015-10-01T03:51:31.734622736Z" level=debug msg="Start untar layer"
Oct 01 03:51:36 docker.vm dockerd[1438]: time="2015-10-01T03:51:36.081522580Z" level=debug msg="Untar time: 4.346893513s"
```

Trying to do the pull direct from the CoreOS VM running docker has the same hanging affect.

```
Expected Results:

johnt@newjim ~⚡ docker pull ubuntu                                                                                                                                                                                                                                        
latest: Pulling from library/ubuntu

d3a1f33e8a5a: Pull complete
c22013c84729: Pull complete
d74508fb6632: Pull complete
91e54dfb1179: Pull complete
library/ubuntu:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.

Digest: sha256:73fbe2308f5f5cb6e343425831b8ab44f10bbd77070ecdfbe4081daa4dbe3ed1
Status: Downloaded newer image for ubuntu:latest
johnt@newjim ~⚡
```

```
Additional Info:

Works on my OSX 10.10.5 laptop with the same vmware fusion and CoreOS versions.
However my docker client on the laptop is Docker version 1.7.1, build 786b29.
(Darwin newjim.local 14.5.0 Darwin Kernel Version 14.5.0: Wed Jul 29 02:26:53 PDT 2015; root:xnu-2782.40.9~1/RELEASE_X86_64 x86_64)

Docker version:
                                                                                                                                                                                                                                            
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): darwin/amd64
Server version: 1.8.2
Server API version: 1.20
Go version (server): go1.4.2
Git commit (server): fae4436-dirty
OS/Arch (server): linux/amd64
```

Seems related to https://github.com/docker/docker/pull/15545 and https://github.com/docker/docker/issues/12823.

Thoughts?

**Comment by johnt337** [comment]
So I figured out the issue.

I found this error in the logs:

```
$ sudo journalctl -f --unit=docker.service
Oct 01 05:02:26 docker.vm dockerd[700]: time="2015-10-01T05:02:26.358906811Z" level=debug msg="Error trying v2 registry: symlink DigiCert_High_Assurance_EV_Root_CA.pem /var/lib/docker/overlay/918fa13f3a5411a37e00692425213a1b01e5ee8a6919f9d32604112f0e5f9dc9/tmproot529186420/etc/ssl/certs/81b9768f.0: no space left on device"
```

I checked how much disk was free for /var/lib/docker

```
$ df -h /var/lib/docker
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdb                   20G  7.7G   11G  42% /var/lib/docker
```

I didn't find this device hitting 100%, but for giggles, I added 10gb to /var/lib/docker and did a resize2fs.

```
$ sudo resize2fs -p /dev/sdb
resize2fs 1.42.13 (17-May-2015)
Filesystem at /dev/sdb is mounted on /var/lib/docker; on-line resizing required
old_desc_blocks = 2, new_desc_blocks = 2
The filesystem on /dev/sdb is now 7864320 (4k) blocks long.
```

Check the size of /var/lib/docker again

```
$ df -h /var/lib/docker
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdb         30G  8.0G   21G  29% /var/lib/docker
```

Ran another docker pull:

```
ohnt@projim ~/g/p/c/m/d/s/d/m/domain⚡ docker pull ubuntu                                                                                           1:06   ✭ mpsplat-517-icd-public_ips|258656d ‹system›
Using default tag: latest
latest: Pulling from library/ubuntu

d3a1f33e8a5a: Pull complete
c22013c84729: Pull complete
d74508fb6632: Pull complete
91e54dfb1179: Pull complete
library/ubuntu:latest: The image you are pulling has been verified. Important: image verification is a tech preview feature and should not be relied on to provide security.

Digest: sha256:73fbe2308f5f5cb6e343425831b8ab44f10bbd77070ecdfbe4081daa4dbe3ed1
Status: Downloaded newer image for ubuntu:latest
```

Not sure why 300mb would matter when 10gb was available. Is this something else related to the overlay fs?

I tried rebooting several times before hand and that didn't make a difference.

**Comment by hosh** [comment]
I am seeing something similar for 829.0.0, Docker 1.8.2. Manual docker pulls get stuck. I'm using 14G for the root device, which includes `/var/lib/docker` This is all running on AWS, so I'd rather not resize things if I don't have to. I don't see anything about running out of diskspace though. It just hangs. Restarting the docker service seems to release the lock, but that's about it.

**Comment by thaJeztah** [comment]
I think this may be caused by inode exhaustion, which is a known issue with overlayfs; see https://github.com/docker/docker/issues/10613. Also read that thread to see how to check if you're running out of inodes.

**Comment by hosh** [comment]
@thaJeztah I figured out why. I don't know about 815.0.0, but 829.0.0 and later compiles Docker with Go 1.5.1. Apparently, there is already a fix that addresses this issue in Docker 1.9.0-dev (don't remember the specific issue number off-hand); the fix addresses this as a side-effect to fixing a different issue with getting it to work with Go 1.5.

I had tested the inode problem, but this problem persist even with an incredible amount of inodes available.

The workaround I used was to download the Docker binaries directly from the Docker github page itself, drop it into CoreOS, and then bind-mount the binary on top of `/usr/bin/docker` via `/etc/fstab`. I have confirmed that 1.9.0-dev (a few days before the prerelease came out) did fix this issue, but flanneld 0.5.1 stopped working with it for whatever reason. So I'm just using Docker 1.8.3 instead for now.

I figure it's better to just use the officially-compiled version of Docker until 1.9.0 hits the CoreOS release channel.

**Comment by thaJeztah** [comment]
Thanks for that information @hosh. I'm not a CoreOS user myself, so it's appreciated :+1: 
In case you missed; we now have the first release-candidates for 1.9, so if other users want to test if that version resolve the issue, here are the links: https://github.com/docker/docker/pull/17000#issuecomment-147926878

**Comment by hosh** [comment]
For completeness: https://github.com/docker/docker/pull/15545

**Comment by thaJeztah** [comment]
Thanks @hosh just curious, have you been able to test the final 1.9.0 release on CoreOS? There have been various fixes/changes between release candidates related to networking (and some API fixes). Possibly those resolve the flanneld issue.

**Comment by hosh** [comment]
@thaJeztah nope, I don't really have time to do that -- I am about 5 weeks late in delivering a production cluster. In the thread over at https://github.com/coreos/bugs/issues/916 it looks like the CoreOS folks are going to roll back to Go 1.5 for the time being. Said, "magically fixes hanging for Go 1.5" doesn't inspire confidence.

**Comment by calavera** [comment]
I'm very sorry to hear about this issues @hosh. I'm almost positive that your issue is not related to this one though.

Docker has never built an official version with Go 1.5 until Docker 1.9. What you're experiencing is a problem caused by CoreOS' forked 1.8.2 version of Docker, which they built at some point with Go 1.5. The official Docker releases for the 1.8.X series were all build with Go 1.4, we don't change compilers between patched versions :wink:

I understand that you don't have time to test or releases, specially since it looks like your production environment has a due deadline. As a rule of thumb, I recommend you to not use binaries built nightly for your production environments. Be wary if you see the suffix `-dirty` in a Docker version, we never release official releases with that suffix, we use it only for development versions.

**Comment by hosh** [comment]
@calavera I'm aware of that (I mean, I said it right here: https://github.com/docker/docker/issues/16689#issuecomment-148589400) I have already talked to the CoreOS guys. Your comments are late to the party, and while I appreciate that you are trying to help (I've done my share of support tickets, so I know where that "I am sorry you have this experience" comes from), the recommendation for not using nightly builds in production is not helpful.

Here is where you can be more helpful: I want to make sure I hear this right --  Docker 1.9 is officially supported on Go 1.5, yes? I say this because the CoreOS team is now rolling back to 1.8.x compiled with Go 1.4. I had referenced them PRs/issues from the Docker side of things which fixes this issue in Docker 1.9 on Go 1.5, but now they weren't confident that this hanging issue is fixed for 1.9 on 1.5. If I can get an official word on that here, I can link them to this.

**Comment by LK4D4** [comment]
@hosh Nope, 1.9 is compiled with 1.4.3.
